python-frontend/src/main/java/org/sonar/python/semantic/v2/types/AstBasedPropagation.java

```diff
@@ -0,0 +1,184 @@
+  private final Propagator propagator;
+
+  public AstBasedPropagation(Map<SymbolV2, Set<Propagation>> propagationsByLhs, TypeTable typeTable) {
+    this.propagationsByLhs = propagationsByLhs;
+    this.propagator = new Propagator(typeTable);
+  }
+
+  public Map<SymbolV2, Set<PythonType>> processPropagations(Set<SymbolV2> trackedVars) {
+    computePropagationDependencies(trackedVars);
+
+    Set<SymbolV2> initializedVars = new HashSet<>();
+    Set<Propagation> propagations = getTrackedPropagation(trackedVars);
+
+    applyPropagations(propagations, initializedVars, true);
+    applyPropagations(propagations, initializedVars, false);
+    return propagations.stream().collect(Collectors.groupingBy(Propagation::lhsSymbol, Collectors.mapping(Propagation::rhsType, Collectors.toSet())));
+  }
+
+  private void computePropagationDependencies(Set<SymbolV2> trackedVars) {
+    propagationsByLhs.forEach((lhs, props) -> {
+      if (trackedVars.contains(lhs)) {
+        props.stream()
+          .filter(Assignment.class::isInstance)
+          .map(Assignment.class::cast)
+          .forEach(assignment -> computeDependencies(assignment, trackedVars));
+      }
+    });
+  }
+
+  private void computeDependencies(Assignment assignment, Set<SymbolV2> trackedVars) {
+    Deque<Expression> workList = new ArrayDeque<>();
+    workList.push(assignment.rhs());
+    while (!workList.isEmpty()) {
+      Expression e = workList.pop();
+      if (e instanceof Name name) {
+        SymbolV2 symbol = name.symbolV2();
+        if (symbol != null && trackedVars.contains(symbol)) {
+          assignment.addVariableDependency(symbol);
+          propagationsByLhs.get(symbol).forEach(propagation -> propagation.addDependent(assignment));
+        }
+      } else if (e instanceof HasTypeDependencies hasTypeDependencies) {
+        workList.addAll(hasTypeDependencies.typeDependencies());
+      }
+    }
+  }
+
+  private Set<Propagation> getTrackedPropagation(Set<SymbolV2> trackedVars) {
+    Set<Propagation> trackedPropagations = new HashSet<>();
+    propagationsByLhs.forEach((lhs, propagations) -> {
+      if (trackedVars.contains(lhs)) {
+        trackedPropagations.addAll(propagations);
+      }
+    });
+    return trackedPropagations;
+  }
+
+  private void applyPropagations(Set<Propagation> propagations, Set<SymbolV2> initializedVars, boolean checkDependenciesReadiness) {
+    Set<Propagation> workSet = new HashSet<>(propagations);
+    while (!workSet.isEmpty()) {
+      Iterator<Propagation> iterator = workSet.iterator();
+      Propagation propagation = iterator.next();
+      iterator.remove();
+      if (!checkDependenciesReadiness || propagation.areDependenciesReady(initializedVars)) {
+        boolean learnt = propagator.propagate(propagation, initializedVars);
+        if (learnt) {
+          workSet.addAll(propagation.dependents());
+        }
+      }
+    }
+  }
+
+
+  private record Propagator(TypeTable typeTable) {
```

Comment:
why do we need this nested record? It looks like we could simply move its methods to the `AstBasedPropagation` and that's it

Reply:
That reason I moved it to a separate record is to show that it is separate from the rest of the class. I feel it helps with readability, since, as the reader, when you're reading Propagator you can forget about `AstBasedPropagation`. So it helps with compartmentalization (and least in my opinion)